yt_dlp/extractor/tubetugraz.py

```diff
@@ -21,17 +21,43 @@ def _perform_login(self, username, password):
         if not urlh:
             return
 
-        urlh = self._request_webpage(
+        content, urlh = self._download_webpage_handle(
             urlh.geturl(), None, fatal=False, headers={'referer': urlh.geturl()},
             note='logging in', errnote='unable to log in', data=urlencode_postdata({
                 'lang': 'de',
                 '_eventId_proceed': '',
                 'j_username': username,
                 'j_password': password
             }))
+        if not urlh:
+            return
+
+        if urlh.geturl() == 'https://tube.tugraz.at/paella/ui/index.html':
+            return
 
-        if urlh and urlh.geturl() != 'https://tube.tugraz.at/paella/ui/index.html':
+        needs_tfa = self._html_search_regex(
+            r"<p\b[^>]*>(Bitte geben Sie einen OTP-Wert ein:)</p>", content, "TFA prompt",
+            default=None) is not None
+        if not needs_tfa:
             self.report_warning('unable to login: incorrect password')
+            return
+
+        tfa_token = self._get_tfa_info()
+        content, urlh = self._download_webpage_handle(
+            urlh.geturl(), None, fatal=False, headers={'referer': urlh.geturl()},
+            note='logging in with TFA', errnote='unable to log in with TFA',
+            data=urlencode_postdata({
+                'lang': 'de',
+                '_eventId_proceed': '',
+                'j_tokenNumber': tfa_token,
+            }))
+        if not urlh:
+            return
+
+        if urlh.geturl() == 'https://tube.tugraz.at/paella/ui/index.html':
+            return
+
+        self.report_warning('unable to login: incorrect TFA code')
```

Comment:
This doesnt make sense to me. If the call fails `not urlh`, then give no warning?

Reply:
The call to `self._download_webpage_handle()` will already output an error or warning, so I do not print a separate warning in order to not duplicate the warning.

This behaviour is actually consistent with the previously released version, which also did not output an extra warning when logging in fails.

Reply:
I would also suggest treating this the same as the previous ones that you already reviewed and combine the two returning if branches:

```suggestion
        if not urlh or urlh.geturl() == 'https://tube.tugraz.at/paella/ui/index.html':
            return

        self.report_warning('unable to login: incorrect TFA code')
```

Reply:
applied this change in my latest commit. Resolving this conversation for now.